### ISSUE TITLE
[FEATURE] - CustomSource implementation 

### ISSUE THREAD

**Issue by Alexboiboi** [magpylib/magpylib#409]
Users should be able to implement their own custom source by defining a H and/or B-field functions. The idea behind it is to implement a source object originating from interpolated FEM or measurement data which can be treated as any classical object in the magpylib library, for instance, object rotation and displacement. The provided field function should only return a (n,3) array in the local object coordinate system, since the library takes care of orientation and displacement.

Sou
- [x] create CustomSource class
- [x] accepts `field_B_lambda` and `field_H_lambda` functions
- [x] validate inputs
- [x] docstring
- [x] docstring example
- [x] plotly representation
- [x] matplotlib representation
- [x] write tests

Current work going on at https://github.com/magpylib/magpylib/tree/custom_source

A code example can be added when docs_v4 and this branch are merged (#389)

**Comment by Alexboiboi** [comment]
Advanced example:

```python
import numpy as np
from scipy.interpolate import RegularGridInterpolator
import magpylib as magpy

magpy.defaults.display.backend = "plotly"

# %%
# Define interpolating function
# -----------------------------
def interpolate_field(data, method="linear", bounds_error=False, fill_value=np.nan):
    """ Creates a 3d-vector field interpolation of a rasterized data from a regular grid

    Parameters
    ----------
    data: numpy.ndarray or array-like
        array of shape (n,6). In order to be a regular grid, the first dimension n
        corresponds to the product of the unique values in x,y,z-directions. 
        The second dimension must have the following ordering on the second axis:
            ``x, y, z, field_x, field_y, field_z``

    method : str, optional
        The method of interpolation to perform. Supported are "linear" and
        "nearest". This parameter will become the default for the object's
        ``__call__`` method. Default is "linear".

    bounds_error : bool, optional
        If True, when interpolated values are requested outside of the
        domain of the input data, a ValueError is raised.
        If False, then `fill_value` is used.

    fill_value : number, optional
        If provided, the value to use for points outside of the
        interpolation domain. If None, values outside
        the domain are extrapolated.

    Returns
    -------
        callable: interpolating function for field values
    """
    data = np.array(data)
    idx = np.lexsort((data[:, 2], data[:, 1], data[:, 0]))  # sort data by x,y,z
    x, y, z, *field_vec = data[idx].T
    X, Y, Z = np.unique(x), np.unique(y), np.unique(z)
    nx, ny, nz = len(X), len(Y), len(Z)
    kwargs = dict(bounds_error=bounds_error, fill_value=fill_value, method=method)
    field_interp = []
    for k, kn in zip((X, Y, Z), "xyz"):
        assert (
            np.unique(np.diff(k)).shape[0] == 1
        ), f"not a regular grid in {kn}-direction"
    for field in field_vec:
        rgi = RegularGridInterpolator((X, Y, Z), field.reshape(nx, ny, nz), **kwargs)
        field_interp.append(rgi)
    return lambda x: np.array([field(x) for field in field_interp]).T


# %%
# Create virtual measured data
# ----------------------------

# create a source
cube = magpy.magnet.Cuboid(
    magnetization=(0, 0, 1000), position=(-10, 0, 0), dimension=(10, 10, 10)
)
dim = [4, 4, 4]
Nelem = [2, 2, 2]
slices = [slice(-d / 2, d / 2, N * 1j) for d, N in zip(dim, Nelem)]
positions = np.mgrid[slices].reshape(len(slices), -1).T

# %%
# get data from a regular grid of positions
Bcube = cube.getB(positions)
# Bcube *= 1 + np.random.random_sample(B.shape)*0.01 # add 1% white noise
Bdata = np.hstack([positions, Bcube])

# %%
# check the interpolation
field_B_lambda = interpolate_field(Bdata)
display("Check field function vs magpylib Cuboid", Bcube, field_B_lambda(positions))

# %%
# create custom source with interpolation field
interp_cube = magpy.misc.CustomSource(field_B_lambda=field_B_lambda)

# add a graphical representation to the custom object, in this case a transparent cube
mesh_data = {
    "type": "mesh3d",
    "i": np.array([7, 0, 0, 0, 4, 4, 2, 6, 4, 0, 3, 7]),
    "j": np.array([0, 7, 1, 2, 6, 7, 1, 2, 5, 5, 2, 2]),
    "k": np.array([3, 4, 2, 3, 5, 6, 5, 5, 0, 1, 7, 6]),
    "x": np.array([-1.0, -1.0, 1.0, 1.0, -1.0, -1.0, 1.0, 1.0]) * 0.5 * dim[0],
    "y": np.array([-1.0, 1.0, 1.0, -1.0, -1.0, 1.0, 1.0, -1.0]) * 0.5 * dim[1],
    "z": np.array([-1.0, -1.0, -1.0, -1.0, 1.0, 1.0, 1.0, 1.0]) * 0.5 * dim[2],
    "opacity": 0.2,
}
interp_cube.style.mesh3d = dict(data=mesh_data, show=True, replace=True)
interp_cube.style.name = 'Interpolated cuboid field'

# %%
# Testing the accuracy of the interpolation
# -----------------------------------------
# .. note::
#    if ``getB`` gets called for positions outside the interpolated field boudaries, the
#    interpolation function will return ``np.nan``. Note that the edges of the domain are
#    susceptible to floating point errors when manipulating an object by rotation, avoid calling
#    positions exactly on the interpolation boundaries since they may yield ``np.nan`` values.

# %%
# define a sensor inside the interpolation boundaries
sens = magpy.Sensor(pixel=positions * 0.5)

# %%
# rotate all object by a common random rotation with common anchor
rotation = dict(
    angle=35, axis=(1, 5, 0.8), anchor=(1, 80, -4)
)  # random rotation parameters
interp_cube.rotate_from_angax(**rotation)
sens.rotate_from_angax(**rotation)
cube.rotate_from_angax(**rotation)

# %%
# display system
magpy.display(cube, sens, interp_cube)

# %%
# compare the interpolated field with the original source
Bcube = sens.getB(cube)
Binterp = sens.getB(interp_cube)
print("Field interpolation error [%]:\n", ((Bcube - Binterp) / Bcube * 100).round(3))

# .. note::
#    The interpolation performance can be in fact arbitrary precise and in this example only 2
#    points per dimension, so that print outputs can be easily shown entirely.
```

**Comment by Alexboiboi** [comment]
closed by #411
